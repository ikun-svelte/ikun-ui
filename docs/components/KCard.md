---
title: KCard
lang: en-US
---

# KCard

Integrate information in a card container.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/card
```

```bash [yarn]
yarn add @ikun-ui/card
```

```bash [npm]
npm install @ikun-ui/card
```

:::

## Basic usage

Card includes title, content.

Card is made up of `header` and `body`. `header` is optional, and its content distribution depends on `header` slot or `title` attribute.

<demo src="card/basic.svelte" github='Card'></demo>

## Simple card

The header part can be omitted.

<demo src="card/simple-card.svelte" github='Card'></demo>

## Shadow

You can define when to show the card shadows

The `shadow` attribute determines when the card shadows are displayed. It can be `always`, `hover` or `never`.

<demo src="card/shadow.svelte" github='Card'></demo>

## Card Props

| Name   | Type                       | Default  | Description                |
| ------ | -------------------------- | -------- | -------------------------- |
| title  | `string`                   | `-`      | title of the card.         |
| shadow | `always \| hover \| never` | `always` | when to show card shadows. |
| cls    | `string`                   | `''`     | Additional class for       |
| attrs  | `Record<string, string>`   | `{}`     | Additional attributes      |

## Card Slots

| Name    | Description             |
| ------- | ----------------------- |
| header  | content of the header.  |
| default | content of the content. |
