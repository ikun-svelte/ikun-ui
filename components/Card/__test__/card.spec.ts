import { tick } from 'svelte';
import { afterEach, expect, test, describe, beforeEach } from 'vitest';
import KCard from '../src';
import KCardHeader from './fixture/card.slot.header.test.svelte';
import KCardDefault from './fixture/card.slot.default.test.svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KCard', () => {
	test('props: cls', async () => {
		const instance = new KCard({
			target: host,
			props: {
				cls: 'k-card--test'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-card--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: title', async () => {
		const instance = new KCard({
			target: host,
			props: {
				title: 'IKun UI'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-card--header')).toBeTruthy();
		expect(host.querySelector('.k-card--header')?.textContent).toBe('IKun UI');
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: shadow', async () => {
		const instance = new KCard({
			target: host,
			props: {
				shadow: 'always'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-card--shadow__always')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
		instance.$set({ shadow: 'hover' });
		await tick();
		expect((host as HTMLElement)!.innerHTML.includes('k-card--shadow__hover')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
		instance.$set({ shadow: 'never' });
		await tick();
		expect((host as HTMLElement)!.innerHTML.includes('k-card--shadow__never')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: header', async () => {
		const instance = new KCardHeader({
			target: host
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-card--header-slot')).toBeTruthy();
		expect(host.querySelector('.k-card--header-slot')?.textContent).toBe('IKun UI');
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: default', async () => {
		const instance = new KCardDefault({
			target: host
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-card--default-slot')).toBeTruthy();
		expect(host.querySelector('.k-card--default-slot')?.textContent).toBe('IKun UI');
		expect(host.innerHTML).matchSnapshot();
	});
});
