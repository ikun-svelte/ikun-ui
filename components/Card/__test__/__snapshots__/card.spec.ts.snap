// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KCard > props: cls 1`] = `"<div class=\\"k-card k-card__dark k-card--shadow__always k-card--shadow__always__dark k-card--transition k-card--test\\"> <div class=\\"k-card--body\\"></div></div>"`;

exports[`Test: KCard > props: shadow 1`] = `"<div class=\\"k-card k-card__dark k-card--shadow__always k-card--shadow__always__dark k-card--transition\\"> <div class=\\"k-card--body\\"></div></div>"`;

exports[`Test: KCard > props: shadow 2`] = `"<div class=\\"k-card k-card__dark k-card--shadow__hover k-card--shadow__hover__dark k-card--transition\\"> <div class=\\"k-card--body\\"></div></div>"`;

exports[`Test: KCard > props: shadow 3`] = `"<div class=\\"k-card k-card__dark k-card--shadow__never k-card--shadow__never__dark k-card--transition\\"> <div class=\\"k-card--body\\"></div></div>"`;

exports[`Test: KCard > props: title 1`] = `"<div class=\\"k-card k-card__dark k-card--shadow__always k-card--shadow__always__dark k-card--transition\\"><div class=\\"k-card--header\\">IKun UI</div> <div class=\\"k-card--body\\"></div></div>"`;

exports[`Test: KCard > slots: default 1`] = `"<div class=\\"k-card k-card__dark k-card--shadow__always k-card--shadow__always__dark k-card--transition\\"> <div class=\\"k-card--body\\"><div class=\\"k-card--default-slot\\">IKun UI</div></div></div>"`;

exports[`Test: KCard > slots: header 1`] = `"<div class=\\"k-card k-card__dark k-card--shadow__always k-card--shadow__always__dark k-card--transition\\"><div class=\\"k-card--header\\"><div class=\\"k-card--header-slot\\" slot=\\"header\\">IKun UI</div></div> <div class=\\"k-card--body\\"></div></div>"`;
